src/zcash/JoinSplit.cpp

```diff
@@ -0,0 +1,344 @@
+#include "JoinSplit.hpp"
+#include "prf.h"
+#include "sodium.h"
+
+#include "zerocash/utils/util.h"
+
+#include <memory>
+
+#include <boost/foreach.hpp>
+#include <boost/format.hpp>
+#include <boost/optional.hpp>
+#include <fstream>
+#include "libsnark/common/default_types/r1cs_ppzksnark_pp.hpp"
+#include "libsnark/zk_proof_systems/ppzksnark/r1cs_ppzksnark/r1cs_ppzksnark.hpp"
+#include "libsnark/gadgetlib1/gadgets/hashes/sha256/sha256_gadget.hpp"
+#include "libsnark/gadgetlib1/gadgets/merkle_tree/merkle_tree_check_read_gadget.hpp"
+
+#include "sync.h"
+
+using namespace libsnark;
+
+namespace libzcash {
+
+#include "zcash/circuit/gadget.tcc"
+
+CCriticalSection cs_ParamsIO;
+CCriticalSection cs_InitializeParams;
+
+template<typename T>
+void saveToFile(std::string path, T& obj) {
+    LOCK(cs_ParamsIO);
+
+    std::stringstream ss;
+    ss << obj;
+    std::ofstream fh;
+    fh.open(path, std::ios::binary);
+    ss.rdbuf()->pubseekpos(0, std::ios_base::out);
+    fh << ss.rdbuf();
+    fh.flush();
+    fh.close();
+}
+
+template<typename T>
+void loadFromFile(std::string path, boost::optional<T>& objIn) {
+    LOCK(cs_ParamsIO);
+
+    std::stringstream ss;
+    std::ifstream fh(path, std::ios::binary);
+
+    if(!fh.is_open()) {
+        throw std::runtime_error((boost::format("could not load param file at %s") % path).str());
+    }
+
+    ss << fh.rdbuf();
+    fh.close();
+
+    ss.rdbuf()->pubseekpos(0, std::ios_base::in);
+
+    T obj;
+    ss >> obj;
+
+    objIn = std::move(obj);
+}
+
+template<size_t NumInputs, size_t NumOutputs>
+class JoinSplitCircuit : public JoinSplit<NumInputs, NumOutputs> {
+public:
+    typedef default_r1cs_ppzksnark_pp ppzksnark_ppT;
+    typedef Fr<ppzksnark_ppT> FieldT;
+
+    boost::optional<r1cs_ppzksnark_proving_key<ppzksnark_ppT>> pk;
+    boost::optional<r1cs_ppzksnark_verification_key<ppzksnark_ppT>> vk;
+    boost::optional<std::string> pkPath;
+
+    static void initialize() {
+        LOCK(cs_InitializeParams);
+
+        ppzksnark_ppT::init_public_params();
+    }
+
+    void preloadProvingKey(std::string path) {
+        pkPath = path;
+    }
+
+    void loadProvingKey() {
+        if (!pk) {
+            if (!pkPath) {
+                throw std::runtime_error("proving key path unknown");
+            }
+            loadFromFile(*pkPath, pk);
+        }
+    }
+
+    void saveProvingKey(std::string path) {
+        if (pk) {
+            saveToFile(path, *pk);
+        } else {
+            throw std::runtime_error("cannot save proving key; key doesn't exist");
+        }
+    }
+    void loadVerifyingKey(std::string path) {
+        loadFromFile(path, vk);
+    }
+    void saveVerifyingKey(std::string path) {
+        if (vk) {
+            saveToFile(path, *vk);
+        } else {
+            throw std::runtime_error("cannot save verifying key; key doesn't exist");
+        }
+    }
```

Comment:
I put a "TODO: #789" in `JoinSplit.hpp` so we could talk about this API another time.